src/hb-ot-layout-common.hh

```diff
@@ -2778,15 +2907,104 @@ struct ConditionSet
     return true;
   }
 
-  bool subset (hb_subset_context_t *c) const
+  int keep_with_variations (hb_collect_feature_substitutes_with_var_context_t *c) const
+  {
+    hb_map_t *condition_map = hb_map_create ();
+    if (unlikely (!condition_map)) return -1;
+
+    hb_set_t *cond_set = hb_set_create ();
+    if (unlikely (!cond_set))
+    {
+      hb_map_destroy (condition_map);
+      return -1;
+    }
+
+    unsigned num_kept_cond = 0, cond_idx = 0;
+    for (const auto& offset : conditions)
+    {
+      unsigned ret = (this+offset).keep_with_variations (c, condition_map);
+      // one condition is not met, drop the entire record
+      if (ret == 2)
+      {
+        hb_map_destroy (condition_map);
+        hb_set_destroy (cond_set);
+        return 2;
+      }
+
+      // axis not pinned, keep this condition
+      if (ret == 0)
+      {
+        cond_set->add (cond_idx);
+        num_kept_cond++;
+      }
+      cond_idx++;
+    }
+
+    // all conditions met
+    if (num_kept_cond == 0) return 1;
+   
+    if (c->num_cond_to_record_map->has (num_kept_cond))
+    {
+      //check if condition_set is unique with variations
+      hb_set_t *record_set = c->num_cond_to_record_map->get (num_kept_cond);
+      for (unsigned record_idx : *record_set)
```

Comment:
This is potentially O(N^2) here's an alternate algorithm that should be  O(N) instead:
- Add a hash() and operator== method to ConditionSet
- Then you should be able to have a hb_hashmap<ConditionSet*, int> (note: the value isn't used so can be whatever type).
- In each call to keep_with_variations call has() on the map with a pointer to the current condition set.  Internally map will dereference the condition set pointer so if there's already an equivalent condition set it will return true. If true drop this condition set.
- If keeping the current set add it to the map.

See how "packed_map" in hb_serialize_context_t works, the idea is basically the same.